src/Invio.Immutable/NullablePropertyHandler.cs

```diff
@@ -0,0 +1,119 @@
+        /// <param name="innerPropertyHandler">
+        ///   The inner <see cref="IPropertyHandler" /> implementation used to handle
+        ///   non-nullable property values.
+        /// </param>
+        /// <param name="property">
+        ///   A <see cref="PropertyInfo" /> on a value object that will potentially
+        ///   have its native equality comparison and hash code generation by the
+        ///   inheriting class.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        ///   Thrown when <paramref name="property" /> is null.
+        /// </exception>
+        /// <exception cref="ArgumentNullException">
+        ///   Thrown when <paramref name="innerPropertyHandler" /> is null.
+        /// </exception>
+        internal NullablePropertyHandler(
+            IPropertyHandler innerPropertyHandler,
+            PropertyInfo property) : base(property) {
+            if (innerPropertyHandler == null) {
+                throw new ArgumentNullException(nameof(innerPropertyHandler));
+            }
+
+            this.innerPropertyHandler = innerPropertyHandler;
+        }
+
+        /// <remarks>
+        ///   This override handles nulls and then dispatches to the inner
+        ///   <see cref="IPropertyHandler"/> used to construct this instance.
+        /// </remarks>
+        /// <inheritdoc />
+        public override bool ArePropertyValuesEqual(object leftParent, object rightParent) {
+            if (leftParent == null) {
+                throw new ArgumentNullException(nameof(leftParent));
+            } else if (rightParent == null) {
+                throw new ArgumentNullException(nameof(rightParent));
+            }
+
+            var leftPropertyValue =
+                this.GetPropertyValueImplOrThrow(leftParent, nameof(leftParent));
+            var rightPropertyValue =
+                this.GetPropertyValueImplOrThrow(rightParent, nameof(rightParent));
+
+            // Handle nulls
+
+            if (Object.ReferenceEquals(leftPropertyValue, null)) {
+                return Object.ReferenceEquals(rightPropertyValue, null);
+            } else if (Object.ReferenceEquals(rightPropertyValue, null)) {
+                return false;
+            }
+
+            return this.innerPropertyHandler.ArePropertyValuesEqual(leftParent, rightParent);
+        }
+
+        /// <remarks>
+        ///   This override handles nulls and then dispatches to the inner
+        ///   <see cref="IPropertyHandler"/> used to construct this instance.
+        /// </remarks>
+        /// <inheritdoc />
+        public override int GetPropertyValueHashCode(object parent) {
+            if (parent == null) {
+                throw new ArgumentNullException(nameof(parent));
+            }
+
+            var propertyValue = this.GetPropertyValueImplOrThrow(parent, nameof(parent));
+
+            if (propertyValue == null) {
+                // Arbitrary (but prime!) number for null.
+
+                return 37;
```

Comment:
Maybe this should be exposed as a constant on `PropertyHandlerBase` to ensure its consistency?